index.js

```diff
@@ -0,0 +1,95 @@
+"use strict";
+
+const cookieSession = require("cookie-session");
+const express = require("express");
+const expressHandlebars = require("express-handlebars");
+const githubAuth = require("connect-oauth-github");
+const github = require("github-request");
+const postRequest = require("request").post;
+const Slack = require("./lib/slack");
+const version = require("./package").version;
+const config = require("./config");
+
+const slackClient = new Slack(config.slack);
+
+const app = express();
+app.use(cookieSession({
+	name: "session",
+	keys: config.web.cookieKeys
+}));
+
+app.engine("handlebars", expressHandlebars({ defaultLayout: "main" }));
+app.set("view engine", "handlebars");
+
+const gha = githubAuth.createClient({
+	id: config.github.clientId,
+	secret: config.github.secret,
+	scope: "user:email"
+});
+
+app.get("/", (request, response) => {
+	response.render("index");
+});
+
+app.get("/auth", gha.handshake);
+
+app.get("/signup", gha.authorize, (request, response) => {
+	let ghaUser = gha.users[request.sessionID];
+
+	github.request({
+		path: "/user?access_token=" + ghaUser.accessToken,
+	}, (error, data) => {
+		if (error) {
+			// TODO: Implement error page
+			return response.status(500).render("error");
+		}
+
+		// TODO: Filter on data.login blacklist
+
+		let ageInDays = (Date.now() - new Date(data.created_at)) / (1000 * 60 * 60 * 24);
+		if (ageInDays < config.github.minimumAge) {
+			// TODO: Check whitelist for approved accounts
+		}
+
+		ghaUser.githubProfile = data;
+		response.render("signup", {
+			name: data.name,
+			email: data.email
+		});
+	});
+});
+
+app.post("/signup", gha.authorize, (request, response) => {
+	let ghaUser = gha.users[request.sessionID];
+	let name = ghaUser.githubProfile.name;
+	let email = ghaUser.githubProfile.email;
+
+	slackClient.invite(email, (error, status) => {
+		if (error) {
+			if (error.code === "INVALID_EMAIL") {
+				// TODO: Let the user know
+			}
+
+			return response.status(500).render("error");
+		}
+
+		// Remove the user data to free up memory
+		delete gha.users[request.sessionID];
+
+		if (status === "ALREADY_INVITED" || status === "ALREADY_IN_TEAM") {
+			// TODO: Let the user know they're already invited or already in the team
+		}
+
+		// TODO: Log the GitHub -> email mapping
```

Comment:
Again, this is based on the same statement from above:

> The main reason I want to tie to GitHub is because it gives us a more holistic ability to moderate. In addition to banning from slack, we can also ban on GitHub

I'm not sure how we'd track that without this list. We can use GitHub's search API to find users by email address, but that won't work with private emails (which we presumably get by using the `user:email` scope with GitHub OAuth).

On the other hand, I did say:

> We can do that, though the user can also change their email address in Slack at any time, so our map can get stale.

I had hoped we'd be able to work around that:

> I wonder if we get back a user ID or some other identifier when we invite someone.

But we don't get any data at all in the response from Slack, just the status of the request.